cpu/sam0_common/periph/pwm.c

```diff
@@ -135,6 +178,94 @@ static void poweroff(pwm_t dev)
 #endif
 }
 
+static void _tc_init(Tc *tc, pwm_mode_t mode, uint8_t prescaler, uint8_t res)
+{
+    /* reset TC module */
+    tc->COUNT8.CTRLA.bit.SWRST = 1;
+    while (tc->COUNT8.CTRLA.bit.SWRST) {}
+
+    /* set PWM mode */
+    switch (mode) {
+        case PWM_LEFT:
+            tc->COUNT8.CTRLBCLR.reg = TC_CTRLBCLR_DIR;     /* count up */
+            break;
+        case PWM_RIGHT:
+            tc->COUNT8.CTRLBSET.reg = TC_CTRLBSET_DIR;     /* count down */
+            break;
+        case PWM_CENTER:        /* currently not supported */
+        default:
+            return;
+    }
+
+    /* configure the TC device */
+    tc->COUNT8.CTRLA.reg = TC_CTRLA_PRESCSYNC_GCLK_Val
+                          | TC_CTRLA_PRESCALER(prescaler)
+                          | TC_CTRLA_MODE_COUNT8
+#ifdef TC_CTRLA_WAVEGEN_NPWM
+                          | TC_CTRLA_WAVEGEN_NPWM;
+#else
+                          ;
+
+    /* select the waveform generation mode -> normal PWM */
+    tc->COUNT8.WAVE.reg = (TC_WAVE_WAVEGEN_NPWM);
+#endif
+
+    /* set the selected period */
+    tc->COUNT8.PER.reg = (res - 1);
+
+#ifdef TC_STATUS_SYNCBUSY
+    while (tc->COUNT8.STATUS.bit.SYNCBUSY) {}
+#else
+    while (tc->COUNT8.SYNCBUSY.reg) {}
+#endif
+
+    /* start PWM operation */
+    tc->COUNT8.CTRLA.reg |= TC_CTRLA_ENABLE;
+}
+
+static void _tcc_init(Tcc *tcc, pwm_mode_t mode, uint8_t prescaler, uint16_t res)
+{
+#ifndef REV_TCC
+    (void) tcc;
+    (void) mode;
+    (void) prescaler;
+    (void) res;
+#else
+    /* reset TCC module */
+    tcc->CTRLA.reg = TCC_CTRLA_SWRST;
+    while (tcc->SYNCBUSY.reg & TCC_SYNCBUSY_SWRST) {}
+
+    /* set PWM mode */
+    switch (mode) {
+        case PWM_LEFT:
+            tcc->CTRLBCLR.reg = TCC_CTRLBCLR_DIR;     /* count up */
+            break;
+        case PWM_RIGHT:
+            tcc->CTRLBSET.reg = TCC_CTRLBSET_DIR;     /* count down */
+            break;
+        case PWM_CENTER:        /* currently not supported */
+        default:
+            return;
+    }
+    while (tcc->SYNCBUSY.reg & TCC_SYNCBUSY_CTRLB) {}
+
+    /* configure the TCC device */
+    tcc->CTRLA.reg = TCC_CTRLA_PRESCSYNC_GCLK_Val
+                   | TCC_CTRLA_PRESCALER(prescaler);
+
+    /* select the waveform generation mode -> normal PWM */
+    tcc->WAVE.reg = (TCC_WAVE_WAVEGEN_NPWM);
+    while (tcc->SYNCBUSY.reg & TCC_SYNCBUSY_WAVE) {}
+
+    /* set the selected period */
+    tcc->PER.reg = (res - 1);
+    while (tcc->SYNCBUSY.reg & TCC_SYNCBUSY_PER) {}
+
+    /* start PWM operation */
+    tcc->CTRLA.reg |= TCC_CTRLA_ENABLE;
```

Comment:
Isn't better to remove this in order to only rely on pwm_poweron ?
Just wondering.

Reply:
That would call `poweron()` again - probably wouldn't hurt, but wouldn't be necessary either.

Since the function implicitly disables PWM by writing the `CTRLA` register in the beginning, I think it's just appropriate to enable it the same way.

Reply:
Fine by me.